src/request.rs

```diff
@@ -368,32 +453,90 @@ impl Request {
         &self.method
     }
 
-    /// Returns all query parameters as a vector of key-value pairs.
+    /// Get the parsed url that will be used for this request.
+    ///
+    /// The url might differ from that originally provided when constructing the
+    /// request if additional query parameters have been added using [`Request::query()`].
+    ///
+    /// The url is wrapped in a `Result` since a common use case is to construct
+    /// the [`Request`] using a `&str` in which case the url is parsed.
     ///
     /// ```
     /// # fn main() -> Result<(), ureq::Error> {
     /// # ureq::is_test(true);
     /// let req = ureq::get("http://httpbin.org/get")
-    ///     .query("foo", "42")
-    ///     .query("foo", "43");
+    ///     .query("foo", "bar");
     ///
-    /// assert_eq!(req.query_params(), vec![
-    ///     ("foo".to_string(), "42".to_string()),
-    ///     ("foo".to_string(), "43".to_string())
-    /// ]);
+    /// assert_eq!(req.parsed_url().unwrap().host(), "httpbin.org");
     /// # Ok(())
     /// # }
     /// ```
-    pub fn query_params(&self) -> Vec<(String, String)> {
-        let mut ret = vec![];
+    pub fn parsed_url(&self) -> Result<&ParsedUrl> {
```

Comment:
On the other hand, here, I think it makes sense to return an owned `Result<ParsedUrl>`.

Reply:
However that forces us to clone. Is that really to prefer?

Reply:
Done